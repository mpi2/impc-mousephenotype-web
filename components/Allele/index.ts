export { default as AlleleMap } from './AlleleMap';
export { default as Crispr } from './Crispr';
export { default as ESCell } from './ESCell';
export { default as IVP } from './IVP';
export { default as Mice } from './Mice';
export { default as QCModal } from './QCModal';
export { default as TVP } from './TVP';
