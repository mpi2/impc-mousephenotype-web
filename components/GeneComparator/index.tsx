import { useRouter } from "next/router";
import React, { useEffect, useState } from "react";
import { Button } from "react-bootstrap";

const GeneComparatorContext = React.createContext(null);

export const GeneComparatorProvider = ({ children }) => {
  const [genes, setGenes] = useState([]);

  const addGene = (gene) => {
    if (gene && !genes.includes(gene)) {
      const newGenes = [...genes, gene];
      setGenes(newGenes);
      return newGenes;
    }
    return genes;
  };

  const removeGene = (gene) => {
    if (gene && genes.includes(gene)) {
      genes.splice(genes.indexOf(gene));
      setGenes(genes);
    }
  };

  const resetGenes = (genes: string[]) => {
    setGenes(genes);
  };

  return (
    <GeneComparatorContext.Provider
      value={{ genes, addGene, removeGene, resetGenes }}
    >
      {children}
    </GeneComparatorContext.Provider>
  );
};

export const useGeneComparator = () => React.useContext(GeneComparatorContext);

export const GeneComparatorTrigger = ({ current }: { current?: string }) => {
  const { addGene, genes } = useGeneComparator();
  const router = useRouter();
  return (
    <Button
      style={{
        position: "fixed",
        zIndex: 100,
        bottom: genes.length > 0 || current ? 0 : -100,
        right: 100,
        transition: ".3s ease transform",
      }}
      variant="primary"
      onClick={() => {
        if (current) {
          const updatedGenes = addGene(current);
          return router.push(`/data/compareGenes/?genes=${updatedGenes.join("_")}`);
        }
        router.push(`/data/compareGenes/?genes=${genes.join("_")}`);
      }}
    >
      Compare Genes{" "}
      <span
        className="bg-white"
        style={{
          display: "inline-block",
          height: 22,
          padding: "0 7px",
          borderRadius: 11,
          marginLeft: 6,
          fontWeight: "bold",
          color: '#000',
        }}
      >
        {genes.length}
      </span>
    </Button>
  );
};
