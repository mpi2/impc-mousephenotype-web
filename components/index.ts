export { default as Card } from './Card';
export { default as Search } from './Search';
export { default as AlleleSymbol } from './AlleleSymbol';
export { default as DownloadData } from './DownloadData'