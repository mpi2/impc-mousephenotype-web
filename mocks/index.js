export async function initMocks() {
  if (typeof window === "undefined") {
    const { server } = await import("./server");
    server.listen();
  } else {
    const { worker } = await import("./browser");
    return worker.start({
      serviceWorker: {
        url: "/data/mockServiceWorker.js",
        options: {
          scope: "/",
        }
      }
    });
  }
}

