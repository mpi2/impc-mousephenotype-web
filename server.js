const { createServer } = require('http');
const { parse } = require('url');
const next = require('next');

const dev = process.env.NODE_ENV !== 'production';
const hostname = 'localhost';
const port = 3000;
const app = next({ dev, hostname, port });
const handle = app.getRequestHandler();

app.prepare()
.then(() => {
    createServer(async (req, res) => {
        try {
            const parsedUrl = parse(req.url, true);
            const { pathname, query} = parsedUrl;
            const host = req.headers.host;
            console.log({ host, pathname });
            res.setHeader('X-IMPC-TEST', "true");
            if (host.includes('nginx.mousephenotype-dev.org')) {
                const modifiedPathName = pathname.replace("/data/", "");
                console.log({ modifiedPathName });
                await app.render(req, res, modifiedPathName, query);
                res.setHeader('X-PATH', JSON.stringify({ host, pathname, modifiedPathName }));
            } else {
                await handle(req, res, parsedUrl);
            }

        } catch (err) {
            console.error('Error occurred handling', req.url, err)
            res.statusCode = 500
            res.end('internal server error')
        }
    })
        .once('error', (err) => {
            console.error(err);
            process.exit(1);
        })
        .listen(port, () => {
            console.log(`Ready on http://${hostname}:${port}`);
        })
});