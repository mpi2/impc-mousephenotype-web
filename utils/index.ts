import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};
