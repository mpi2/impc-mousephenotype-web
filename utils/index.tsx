import React from "react";
import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};

export const formatPValue = (pValue: number) => {
  const pValueArray = Number.parseFloat(String(pValue))
    .toExponential(2)
    .split("e");
  return (
    <>
      {pValueArray[0]}x10<sup>{pValueArray[1].replace("+", "")}</sup>
    </>
  );
};

export const formatESCellName = (src: string) => {
  const [name, superscript, end] = src.split(/<|>/);
  return (
    <>
      {name}
      <sup>{superscript}</sup>
      {end}
    </>
  );
};

const fetchCache = {};
export { fetchCache };

export const toSentenceCase = (camelCase) => {
  if (camelCase) {
    const result = camelCase.replace(/([A-Z])/g, " $1");
    return result[0].toUpperCase() + result.substring(1).toLowerCase();
  }
  return "";
};
