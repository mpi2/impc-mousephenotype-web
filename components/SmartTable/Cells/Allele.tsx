import { Model, TableCellProps } from "@/models";
import _ from "lodash";
import { formatAlleleSymbol } from "@/utils";

const AlleleCell = <T extends Model>(props: TableCellProps<T>) => {
  const allele = formatAlleleSymbol(_.get(props.value, props.field) as string);
  return (
    <>
      {allele[0]}
      <sup>{allele[1]}</sup>
    </>
  )
};

export default AlleleCell;