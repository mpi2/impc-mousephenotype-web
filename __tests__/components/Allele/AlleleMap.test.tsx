import { render, screen, waitFor } from '@testing-library/react';
import { AlleleMap } from "@/components/Allele";
import { QueryClientProvider } from "@tanstack/react-query";
import { API_URL, createTestQueryClient } from "../../utils";
import mockRouter from "next-router-mock";
import { GeneComparatorProvider } from "@/components/GeneComparator";
import { server } from "../../../mocks/server";
import { rest } from "msw";
import dbn1AlelleMap from "../../../mocks/data/tests/dbn1-allele-map.json";

jest.mock('next/router', () => jest.requireActual('next-router-mock'));

describe('Allele Page - Allele Map', () => {
  it('renders correctly', async () => {
    const endpointURL = `${API_URL}/api/v1/alleles/es_cell/get_by_mgi_and_allele_name/MGI:1931838/tm1a(KOMP)Wtsi`;
    server.use(
      rest.get(endpointURL, (req, res, ctx) => {
        return res(ctx.status(200), ctx.json(dbn1AlelleMap));
      })
    );

    const client = createTestQueryClient();
    await mockRouter.push('/data/alleles/MGI:1931838/tm1a(KOMP)Wtsi');

    const { container } = render(
      <QueryClientProvider client={client}>
        <GeneComparatorProvider>
          <AlleleMap
            mgiGeneAccessionId="MGI:1931838"
            alleleName="tm1a(KOMP)Wtsi"
            emsembleUrl="https://www.ensembl.org/Mus_musculus/Location/View?g=ENSMUSG00000034675"
          />
        </GeneComparatorProvider>
      </QueryClientProvider>
    );
    await waitFor(() => expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent('Allele Map'))
    expect(container).toMatchSnapshot();
  });

  it('displays nothing if data is empty', async () => {
    const endpointURL = `${API_URL}/api/v1/alleles/es_cell/get_by_mgi_and_allele_name/MGI:1931838/tm1a(KOMP)WtsiTest`;
    server.use(
      rest.get(endpointURL, (req, res, ctx) => {
        return res(ctx.status(200), ctx.json([]));
      })
    );

    const client = createTestQueryClient();
    await mockRouter.push('/data/alleles/MGI:1931838/tm1a(KOMP)WtsiTest');

    const { container } = render(
      <QueryClientProvider client={client}>
        <GeneComparatorProvider>
          <AlleleMap
            mgiGeneAccessionId="MGI:1931838"
            alleleName="tm1a(KOMP)Test"
            emsembleUrl="https://www.ensembl.org/Mus_musculus/Location/View?g=ENSMUSG00000034675"
          />
        </GeneComparatorProvider>
      </QueryClientProvider>
    );
    await waitFor(() => expect(container).toBeEmptyDOMElement());
  });
});