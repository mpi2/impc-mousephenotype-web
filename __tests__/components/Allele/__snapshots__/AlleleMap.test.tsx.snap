// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Allele Page - Allele Map renders correctly 1`] = `
<div>
  <div
    class="card"
  >
    <h2>
      Allele Map
    </h2>
    <p
      class="mb-0"
    >
      <a
        class="primary link"
        href="https://escell-data.s3.eu-west-2.amazonaws.com/targ_rep_allele/11509/escell-clone-genbank-file"
        target="_blank"
      >
        Genbank
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-up-right-from-square "
          data-icon="up-right-from-square"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="margin-left: 5px;"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M352 0c-12.9 0-24.6 7.8-29.6 19.8s-2.2 25.7 6.9 34.9L370.7 96 201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L416 141.3l41.4 41.4c9.2 9.2 22.9 11.9 34.9 6.9s19.8-16.6 19.8-29.6V32c0-17.7-14.3-32-32-32H352zM80 32C35.8 32 0 67.8 0 112V432c0 44.2 35.8 80 80 80H400c44.2 0 80-35.8 80-80V320c0-17.7-14.3-32-32-32s-32 14.3-32 32V432c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V112c0-8.8 7.2-16 16-16H192c17.7 0 32-14.3 32-32s-14.3-32-32-32H80z"
            fill="currentColor"
          />
        </svg>
      </a>
      Â 
      <span
        class="grey ms-2 me-2"
      >
        |
      </span>
      <a
        class="primary link"
        href="https://www.ensembl.org/Mus_musculus/Location/View?g=ENSMUSG00000034675"
        target="_blank"
      >
        Ensembl
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-up-right-from-square "
          data-icon="up-right-from-square"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="margin-left: 5px;"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M352 0c-12.9 0-24.6 7.8-29.6 19.8s-2.2 25.7 6.9 34.9L370.7 96 201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L416 141.3l41.4 41.4c9.2 9.2 22.9 11.9 34.9 6.9s19.8-16.6 19.8-29.6V32c0-17.7-14.3-32-32-32H352zM80 32C35.8 32 0 67.8 0 112V432c0 44.2 35.8 80 80 80H400c44.2 0 80-35.8 80-80V320c0-17.7-14.3-32-32-32s-32 14.3-32 32V432c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V112c0-8.8 7.2-16 16-16H192c17.7 0 32-14.3 32-32s-14.3-32-32-32H80z"
            fill="currentColor"
          />
        </svg>
      </a>
    </p>
    <div>
      <img
        src="https://escell-data.s3.eu-west-2.amazonaws.com/targ_rep_allele/11509/simple.png"
        style="display: block; max-width: 100%;"
      />
    </div>
  </div>
</div>
`;
